### ISSUE TITLE
Potentially incorrect assumpion that network output size == label size.

### ISSUE THREAD

**Issue by jarney** [deeplearning4j/deeplearning4j#3468]
#### Issue Description

In developing a custom loss function based on Bishop's mixture density networks (http://publications.aston.ac.uk/373/1/NCRG_94_004.pdf), I ran across a problem during testing.  In particular, Bishop's loss function requires that the number of outputs of the network be different than the label size because the network is trained to generate a mixture of gaussian distributions for each label where the network output is of size (labelSize * mixturesPerLabel * 3) whereas the size of the labels is simply (labelSize).

Please describe our issue, along with:
- BaseOutputLayer should permit use of loss functions with a potentially different network output size than label size.
- Exception thrown in BaseOutputLayer due to (deliberate) size mismatch:
Exception in thread "main" org.deeplearning4j.exception.DL4JInvalidInputException: Labels array numColumns (size(1) = 2) does not match output layer number of outputs (nOut = 6)
	at org.deeplearning4j.nn.layers.BaseOutputLayer.getGradientsAndDelta(BaseOutputLayer.java:173)
	at org.deeplearning4j.nn.layers.BaseOutputLayer.backpropGradient(BaseOutputLayer.java:151)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(MultiLayerNetwork.java:1097)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.backprop(MultiLayerNetwork.java:1049)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:1833)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:151)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.optimize(BaseOptimizer.java:179)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:51)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1443)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1408)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1487)
	at org.ensor.fftmusings.mdn.example.MixtureDensityNetwork.main(MixtureDensityNetwork.java:112)

I would suggest that the assertion here is not actually necessary.  If there is agreement, I would be happy to raise a pull-request.
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/BaseOutputLayer.java#L176

I have worked around the issue with a custom output layer which overrides this behavior, so this is not a blocking issue, but I feel somewhat uncomfortable overriding some methods at such a core level.

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version : 0.7.2
* platform information (OS, etc) Linux orac 4.8.0-52-generic #55~16.04.1-Ubuntu SMP Fri Apr 28 14:36:29 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* CUDA version, if used : N/A
* NVIDIA driver version, if in use : N/A

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by AlexDBlack** [comment]
Yes, I agree that it's totally valid in some cases for those to be different.

I wouldn't be comfortable for the error to be simply removed, as in >>90% of cases it is indeed an error.
However, we could enforce this as necessary in each loss function instead of in the output layers.

**Comment by jarney** [comment]
I agree that enforcing this type of condition would be best done at the level of the cost-functions.

**Comment by AlexDBlack** [comment]
As an aside - I'd love to have a mixture density network implementation in DL4J, if you are interested in contributing it when you are done. :)

**Comment by jarney** [comment]
I've raised a pair of pull-requests for nd4j and deeplearning4j which move the assertions from the BaseOutputLayer into the LossFunctions.  Please let me know if you feel this is an appropriate solve.

**Comment by AlexDBlack** [comment]
Fixed in the following 2 PRs:
https://github.com/deeplearning4j/deeplearning4j/pull/3473
https://github.com/deeplearning4j/nd4j/pull/1848
